### DISCUSSION TITLE
RFC: just some thoughts on jobs and how to get some...

### DISCUSSION THREAD

**Discussion by desFritz** [endless-sky/endless-sky#7091] [Ideas]
Hello,

just playing around with editing and creating outfits, jobs and such and would like some feedback to an idea.
(Aside from the fact of adding new categories ;-)

![grafik](https://user-images.githubusercontent.com/109269346/182660611-3242ac6a-9bfd-4043-9aa3-dceb427c2cf0.png)
The idea of this crew segment of the outfitter (maybe one day having "Hire Crew" open a screen like this?) is to enable and
disable certain job types. (Want Bounty Hunter interactions? - Get a Bounty Hunter in your crew. Want some evacuations missions and humanitary help missions? Get a medical bay and hire a doctor. Want some repair/tow missions? Get a repair bay and repair crew... and the jobs will appear.  Since there does not seem to be a filter (yet) for offered jobs.)
```
	on offer
		require "Doctor"
```
So basically a way for the player to enable and disable certain mission types. A bit like the luxury accomodation or prison cell.

**Example: Bounty Hunter**

You hire (purchase) a Bounty Hunter in the outfitter. When he is on your flagship and random triggers, certain dialogs appear and missions/ random jobs can be accessed during "landing", or the "spaceport".

![grafik](https://user-images.githubusercontent.com/109269346/182662667-92c36f1a-ee4a-48c8-9753-a4dfb71a2097.png)

![grafik](https://user-images.githubusercontent.com/109269346/182663184-ef6de94f-94a3-4d2a-8009-72368d77a92a.png)

Accepting creates a generic mission...
![grafik](https://user-images.githubusercontent.com/109269346/182663267-09b606e5-0f9d-46b3-9ffa-bbe3396c9493.png)
![grafik](https://user-images.githubusercontent.com/109269346/182663299-2a027d54-c8ae-4ac8-9bec-864b1d3080da.png)

When departing the planet/station the "on enter" triggers and it continues. Not successfully in that case.
![grafik](https://user-images.githubusercontent.com/109269346/182663757-ef7669a5-db6f-4751-a92a-b30e26fdba2a.png)

> decline: The mission is declined. (This is also useful for creating conversations that appear when you land or enter a spaceport, but that are intended just to provide flavor, not to lead to a mission.)

Since providing flavor sound always nice it states:

> enter [<system>]: your ship enters the given system for the first time since this mission was accepted. If no system is specified, this triggers as soon as your ship takes off from the current planet.

So. Hire a Bounty Hunter, get a non repeating mission with 20 planets "on enter <system>", and the first time you enter that system the bounty hunter can tell you a story about a criminal he once caught there, or how dangerous, corrupt, criminal that system is etc?
And since this is a non repeating mission and "on enter" is supposed to trigger once, he will not repeat his banter.

Random jobs might allow to have "follow ship to pirate leader" aka waypoint mission, or spawn enemies after departure you can eliminate for bounty, or just tipp off the authorities, which might lead to a shooting that could kill your bounty hunter, cause trouble and allow you to steal the cargo of the wanted criminals, while they are getting arrested by police (conversation style).
...

Just some ideas, not sure if it's worth developing further.  Any feedback?

**Comment by MasterOfGrey** [comment]
> not sure if it's worth developing further.

So the short answer is "yes" - I think it is worth developing further.

The longer answer is that I do think it needs some refinement.

1 - We should probably start by breaking anything like this which you add to your ship to provide non-ship capabilities out into its own category. IRL most of this stuff falls under the category of "Social Infrastructure" - prisons (brig), healthcare (dr/med bay), hotels (luxury accom), but I don't think that name works particularly well for a ship.

2 - Secondly we should probably start being able to assign an "alternative attribute" to a given attribute as a child of it.
e.g.:
- Having "Luxury Bunks", with a child attribute of just "Bunks" - allowing you to use your luxury bunks as both.
- Maybe "Secure Bunks" on an outfit can also double with a child attribute of just "Bunks".
- ... but actual "Brig Bunks" cannot be used as regular bunks, only having "Secure Bunks" as their child attribute.

The attributes would be defined in the outfit, not inherent to the attribute type, so each outfit could decide if its type of bunk can be used for multiple things or not.

3 - I don't think we should ever just be hiring *a* special crew member. If they don't require any sort of special supplies or support infrastructure, then practically they can just be handwaved as someone that you would have if you have a suitably large ship that it would be necessary to have one. - Like if your ship has `crew` > 15, you can probably assume there's a qualified engineer on board, if your `cargo space` is > 50 tonnes and your `crew` is greater then 5 you probably have a loadmaster, etc.

4 - If you always combine it with some support equipment and with the code changes from 2 then you really start to get flexibility involved. Like, maybe a "med bed" takes away 1 bunk, adds one crew requirement, and only adds a medic attribute - unlocking only very small missions, but a given size of "med bay" includes at least 1 "doctor" and as many "medics" as there are additional staff added to the crew requirement (you'd probably make a medbay with 5 total crew have 4 medics and 1 doctor with the child attribute of medic, since the doctor would also be capable of handling medic missions, but you wouldn't be able to use 6 medical staff worth of missions still). 

These are just my thoughts though, for developing this further though I would be very interested to hear from @quyykk @petervdmeer @Terin @Zitchas @Galaucus @warp-core @samrocketman - and ultimately Amazinite though I won't tag him just yet so he doesn't get mega-spammed. (Sorry for the mass-ping to the rest of you.)

**Comment by petervdmeer** [comment]
The bounty hunter story (and other stories) look nice, I suggest you proceed with them and create a nice content PR for them. But I suggest to add this bounty-hunter without the "trinket" crew outfit; temporary hiring a bounty-hunter can also be done at a spaceport by using a special mission that triggers in the spaceport. The other mission can then trigger based on the bounty-hunter mission being active (and other conditions related to each-other for ordering them).

(I'm not against new "story related outfits", but if the story works just as well without such outfit, then I suggest to focus on the story.)

Being able to trigger jobs on departure sounds like a feature I also want to have. It is probably similar to the triggering in [de-orbit](https://github.com/endless-sky/endless-sky/issues/6225) and  [outfitters](https://github.com/endless-sky/endless-sky/issues/6652). Allowing filters on attributes and outfits in `to offer` also sounds something I would like to have and some work is in progress to get us close to this.

**Comment by desFritz** [comment]
Agree: de-orbiting, landed, pre-landing, entering etc. would be nice starts for missions.

Running into a minor problem right now.

_**Example mission:**_
Landing on planet (= `landing`)
Conversation starts. (= `on offer`)
Bounty Hunter: "Located a suspicious ship. Shall I investigate"?
Yes (= `accept`), conversation ends
Leaving planet (= `on enter`)
Conversation starts.
Bounty Hunter: "They are ex convicts with no active warrant. We should scan their freighter `npc`, just in case"
npc freighter target spawns
scan freighter
Conversation starts (as subelement of npc)
Bounty Hunter: "Sadly they do not smuggle something. We should let them go... for now."

Problem:
The mission does not get finished. It seems like the mission design for `npc scan cargo` is to return to source or go to destination for the mission to succeed. Maybe for other mission types as well...
Is there a way to abort a mission from conversation? Like accept und decline etc?

With the above mentioned way you have to land once more on the same planet for the mission to succeed and disappear from the list of active missions.

Making it invisible is not an option (bad coding/writing).
FR for changing "npc scan cargo" to succeed after having done the scan without given destination would break existing missions, since they are written without destination, but expect you to return to source planet, or start in local system and require to go somewhere after the scan.

So `abort` as something you can do from a conversation within a mission? Or is there an existing way already? An aborting - option does not sound that 'dirty' to me, but I don't know the code. Have searched the game missions/jobs and the documentary, but so far it seems after the scan you have to go somewhere.

Except for the trail of open/active missions the bounty hunter idea works so far nicely. Even if it feels like working around the system, but most likely I am not used enough to the logic/design of the ES missions yet.

Requested additions in  https://github.com/endless-sky/endless-sky/issues/6225  will be appreciated.

**Comment by RisingLeaf** [comment]
Adding a note here regarding #7406 as requested by @MasterOfGrey.

The PR allows the creation of bunk types that can either carry crew or passengers or both or none of them.
The last one is the one that might be interesting for prisoner transport, a bunk type brick could be created that is not used by crew or passengers and could be checked for in prisoner missions.
Also luxury bunks maybe used by passengers and not by crew?
